File: ios/Packages/SimplicityAdmin/Sources/SimplicityAdmin/Invitations/InvitationsViewModel.swift
import Factory
import Foundation
import SimplicityApi
import SimplicityFoundation
import SimplicityServices

@Observable
@MainActor
public final class InvitationsViewModel {

    // MARK: Dependencies

    @ObservationIgnored @Injected(\.organisationService) private var organisations
    @ObservationIgnored @Injected(\.sessionService) private var session

    // MARK: Properties

    public private(set) var invitations: [InvitationResponse] = []
    public private(set) var teams: [TeamResponse] = []
    public private(set) var isLoading = false
    public private(set) var isBusy = false
    public private(set) var errorMessage: String?

    public var email: String = .empty
    public var orgRole: CreateInvitationRequest.OrgRole = .orgMember
    public var teamId: UUID?
    public var teamRole: CreateInvitationRequest.TeamRole = .teamMember

    /// Client-side email validation is usually not worth doing, but an invitation to a malformed
    /// address fails silently from the sender's point of view: they see it sent and nobody arrives.
    public var canInvite: Bool {
        let trimmed = email.trimmingCharacters(in: .whitespaces)
        return trimmed.contains("@")
            && !trimmed.hasPrefix("@")
            && !trimmed.hasSuffix("@")
            && !isBusy
    }

    private var orgId: UUID?

    // MARK: Init

    public init() {}

    // MARK: Functions

    public func load() async {
        orgId = await session.current?.activeOrganisationId
        guard let orgId else {
            isLoading = false
            return
        }

        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            invitations = try await organisations.invitations(orgId: orgId)
            teams = try await organisations.teams(orgId: orgId)
        } catch {
            errorMessage = String(localized: "invitations_load_failed", bundle: .module)
        }
    }

    public func invite() async {
        guard canInvite, let orgId else { return }

        isBusy = true
        errorMessage = nil
        defer { isBusy = false }

        do {
            let created = try await organisations.invite(
                orgId: orgId,
                email: email,
                orgRole: orgRole,
                teamId: teamId,
                teamRole: teamRole
            )
            invitations.append(created)
            email = .empty
            teamId = nil
        } catch {
            // The address stays in the field so it does not have to be retyped.
            errorMessage = String(localized: "invitations_send_failed", bundle: .module)
        }
    }

    public func revoke(_ invitation: InvitationResponse) async {
        guard let orgId, let id = invitation.id, !isBusy else { return }

        isBusy = true
        errorMessage = nil
        defer { isBusy = false }

        do {
            try await organisations.revokeInvitation(orgId: orgId, invitationId: id)
            invitations.removeAll { $0.id == id }
        } catch {
            errorMessage = String(localized: "invitations_revoke_failed", bundle: .module)
        }
    }
}

public extension InvitationResponse {

    /// Only a pending invitation can be revoked. One already accepted is a membership now, and
    /// removing that person is a different action in a different place.
    var canRevoke: Bool {
        status == .pending
    }
}

File: android/admin/src/main/kotlin/io/simplicity/training/admin/InvitationsViewModel.kt
package io.simplicity.training.admin

import androidx.lifecycle.ViewModel
import io.simplicity.training.api.models.CreateInvitationRequest
import io.simplicity.training.api.models.InvitationResponse
import io.simplicity.training.services.OrganisationService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import java.util.UUID

data class InvitationsState(
    val invitations: List<InvitationResponse> = emptyList(),
    val email: String = "",
    val role: CreateInvitationRequest.OrgRole = CreateInvitationRequest.OrgRole.ORG_MEMBER,
    val teamId: UUID? = null,
    val isBusy: Boolean = false,
    val failed: Boolean = false,
) {
    /**
     * A minimal check rather than a full address grammar. An invitation to a malformed address
     * fails silently from the sender's point of view — they see it sent and nobody arrives — so
     * catching the obvious mistakes is worth it, and rejecting valid oddities is not.
     */
    val canInvite: Boolean
        get() = email.trim().let { it.contains("@") && !it.startsWith("@") && !it.endsWith("@") } && !isBusy
}

class InvitationsViewModel(
    private val organisations: OrganisationService,
    private val orgId: UUID,
) : ViewModel() {

    private val _state = MutableStateFlow(InvitationsState())
    val state: StateFlow<InvitationsState> = _state.asStateFlow()

    fun emailChanged(value: String) = _state.update { it.copy(email = value) }
    fun roleChanged(value: CreateInvitationRequest.OrgRole) = _state.update { it.copy(role = value) }
    fun teamChanged(value: UUID?) = _state.update { it.copy(teamId = value) }

    suspend fun load() {
        _state.update { it.copy(isBusy = true, failed = false) }
        try {
            _state.update { it.copy(invitations = organisations.invitations(orgId), isBusy = false) }
        } catch (e: Exception) {
            _state.update { it.copy(isBusy = false, failed = true) }
        }
    }

    suspend fun invite() {
        val current = _state.value
        if (!current.canInvite) return

        _state.update { it.copy(isBusy = true, failed = false) }
        try {
            val created = organisations.invite(orgId, current.email.trim(), current.role, current.teamId)
            _state.update {
                it.copy(
                    invitations = it.invitations + created,
                    email = "",
                    teamId = null,
                    isBusy = false,
                )
            }
        } catch (e: Exception) {
            // The address stays in the field so it does not have to be retyped.
            _state.update { it.copy(isBusy = false, failed = true) }
        }
    }

    suspend fun revoke(invitation: InvitationResponse) {
        val id = invitation.id ?: return
        _state.update { it.copy(isBusy = true, failed = false) }
        try {
            organisations.revokeInvitation(orgId, id)
            _state.update { it.copy(isBusy = false) }
            load()
        } catch (e: Exception) {
            _state.update { it.copy(isBusy = false, failed = true) }
        }
    }
}
